src/Routing/PushState.purs

```diff
@@ -0,0 +1,241 @@
+    obsvNode <-
+      DOM.window
+        >>= DOM.document
+        >>> map DOM.htmlDocumentToDocument
+        >>= DOM.createTextNode ""
+        >>> map DOM.textToNode
+    observer <- DOM.mutationObserver \_ _ -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef []
+      for_ argsQueue notify
+    DOM.observe obsvNode { characterData: true } observer
+    pure \args -> do
+      argsQueue <- readRef argsRef
+      writeRef argsRef (Array.snoc argsQueue args)
+      when (Array.null argsQueue) do
+        fresh ← readRef freshRef
+        writeRef freshRef (fresh + 1)
+        DOM.setNodeValue (show fresh) obsvNode
+
+  let
+    stateFn op state path = do
+      let
+        loc = normalizeLocation state path
+        url = DOM.URL $ unwrap basename <> loc.pathname <> loc.search <> loc.hash
+      DOM.window
+        >>= DOM.history
+        >>= op state (DOM.DocumentTitle "") url
+      schedule loc
+
+    listener =
+      DOM.eventListener \_ -> notify =<< locationState
+
+  DOM.window
+    >>= DOM.windowToEventTarget
+    >>> DOM.addEventListener DOM.popstate listener false
+
+  pure
+    { pushState: stateFn DOM.pushState
+    , replaceState: stateFn DOM.replaceState
+    , locationState
+    , listen
+    }
+
+type LocationState =
+  { state :: Foreign
+  , path :: String
+  , pathname :: String
+  , search :: String
+  , hash :: String
+  }
+
+normalizeLocation :: Foreign -> Path -> LocationState
+normalizeLocation state (Path path) =
+  case searchIx, hashIx of
+    Nothing, Nothing   -> { state, pathname: path, hash: "", search: "", path }
+    Nothing, Just hix  -> { state, pathname: String.take hix path, search: "", hash: String.drop hix path, path }
+    Just six, Nothing  -> { state, pathname: String.take six path, search: String.drop six path, hash: "", path }
+    Just six, Just hix -> { state, pathname: String.take six path, search: String.take (hix - six) (String.drop six path), hash: String.drop hix path, path }
+  where
+  searchIx = String.indexOf (String.Pattern "?") path
+  hashIx = String.indexOf (String.Pattern "#") path
+
+-- | Folds effectfully over location changes given callbacks for handling
+-- | changes and the initial location. Returns an effect which removes the
+-- | listener.
+foldLocations
+  :: forall eff a
+   . (a -> LocationState -> Eff (PushStateEffects eff) a)
+  -> (LocationState -> Eff (PushStateEffects eff) a)
+  -> PushStateInterface (PushStateEffects eff)
+  -> Eff (PushStateEffects eff) (Eff (PushStateEffects eff) Unit)
+foldLocations cb init psi = do
+  ref <- newRef =<< init =<< psi.locationState
+  psi.listen (\loc -> writeRef ref =<< flip cb loc =<< readRef ref)
+
+-- | Runs the callback on every location change providing the previous location
+-- | and the latest location. Returns an effect which removes the listener.
+locations
+  :: forall eff
+   . (Maybe LocationState -> LocationState -> Eff (PushStateEffects eff) Unit)
+  -> PushStateInterface (PushStateEffects eff)
```

Comment:
IIRC, I usually see the callback arg as the last argument. I see you used `flip locations` yourself, so might want to change the arg ordering.

Reply:
I used this argument order just because that's the order of something like `foldM`. It's less convenient for inline callbacks, more convenient for currying, and otherwise benign.